pkg/trie/triedb/triedb_iterator.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package triedb
+
+import (
+	"bytes"
+
+	nibbles "github.com/ChainSafe/gossamer/pkg/trie/codec"
+	"github.com/ChainSafe/gossamer/pkg/trie/triedb/codec"
+)
+
+type iteratorState struct {
+	parentFullKey []byte     // key of the parent node of the actual node
+	node          codec.Node // actual node
+}
+
+// fullKeyNibbles return the full key of the node contained in this state
+// child is the child where the node is stored in the parent node
+func (s *iteratorState) fullKeyNibbles(child *int) []byte {
+	fullKey := bytes.Join([][]byte{s.parentFullKey, s.node.GetPartialKey()}, nil)
+	if child != nil {
+		return bytes.Join([][]byte{fullKey, {byte(*child)}}, nil)
+	}
+
+	return nibbles.NibblesToKeyLE(fullKey)
+}
+
+type TrieDBIterator struct {
+	db        *TrieDB          // trie to iterate over
+	nodeStack []*iteratorState // Pending nodes to visit
+}
+
+func NewTrieDBIterator(trie *TrieDB) *TrieDBIterator {
+	rootNode, err := trie.getRootNode()
+	if err != nil {
+		panic("trying to create trie iterator with incomplete trie DB")
```

Comment:
I think if getting the rootNode fails and then corrupts our db or breaks storage, panic could be good. But I totally agree with @kishansagathiya that we need to make sure to use it only in extreme cases. For me, its pretty much if the correctness of the node is broken and we just need to shutdown.

However if we are going to panic, I think its a good question why not return the err as well in the panic?

Reply:
Good point, I'll change it to return an error. Thanks for the comments!

Reply:
This will make me modify all methods and the trie interface to start handling and returning the new error, I'll leave it for a future PR to prevent adding too many changes here

Reply:
> This will make me modify all methods and the trie interface to start handling and returning the new error, I'll leave it for a future PR to prevent adding too many changes here

If this is the case and if you have used a lot of panics that should otherwise just be errors, let's create an issue if we are going to do as a separate PR. It's okay to tackle this separately, but let's make this the next PR.

Btw, this panic thing is part of list of go proverbs. They seem like good rules to follow https://go-proverbs.github.io/

Reply:
Totally agree! thanks, I'll address it in #3942